sockeye/generate_decoder_states.py

```diff
@@ -0,0 +1,239 @@
+
+        self.output_prefix = output_prefix
+        self.state_store_file = None
+        self.words_store_file = None
+
+        # info for KNNConfig
+        self.num_states = 0
+        self.dimension = None
+        self.state_data_type = utils.get_numpy_dtype(state_data_type)
+        self.word_data_type = utils.get_numpy_dtype(word_data_type)
+
+    @staticmethod
+    def probe_token_count(target: str, max_seq_len: int) -> int:
+        token_count = 0
+        with codecs.open(target, 'r', 'utf-8') as f:
+            for line in f:
+                token_count += min(len(line.strip().split(' ')) + 1, max_seq_len)  # +1 for EOS
+        return token_count
+
+    def init_store_file(self, initial_size: int) -> None:
+        self.dimension = self.model.config.config_decoder.model_size
+
+        self.state_store_file = NumpyMemmapStorage(get_state_store_filename(self.output_prefix),
+                                                  self.dimension, self.state_data_type)
+        self.words_store_file = NumpyMemmapStorage(get_word_store_filename(self.output_prefix),
+                                                  1, self.word_data_type)
+        self.state_store_file.open(initial_size, 1)
+        self.words_store_file.open(initial_size, 1)
+
+    def generate_states_and_store(self,
+                              sources: List[str],
+                              targets: List[str],
+                              batch_size: int) -> None:
+        assert self.state_store_file != None, \
+               "You should call probe_token_count first to initialize the store files."
+
+        # get data iter
+        data_iter = data_io.get_scoring_data_iters(
+            sources=sources,
+            targets=targets,
+            source_vocabs=self.source_vocabs,
+            target_vocabs=self.target_vocabs,
+            batch_size=batch_size,
+            max_seq_len_source=self.max_seq_len_source,
+            max_seq_len_target=self.max_seq_len_target
+        )
+
+        with pt.inference_mode():
+            for batch_no, batch in enumerate(data_iter, 1):
+                if (batch_no + 1) % 1000 == 0:
+                    logger.info("At batch number {0}".format(batch_no + 1))
+    
+                # get decoder states
+                batch = batch.load(self.device)
+                model_inputs = (batch.source, batch.source_length, batch.target, batch.target_length)
+                if self.traced_model is None:
+                    trace_inputs = {'get_decoder_states': model_inputs}
+                    self.traced_model = pt.jit.trace_module(self.model, trace_inputs, strict=False)
+                decoder_states = self.traced_model.get_decoder_states(*model_inputs)  # shape: (batch, sent_len, hidden_dim)
+    
+                # flatten batch and sent_len dimensions, remove pads on the target
+                pad_mask = (batch.target != C.PAD_ID).squeeze(2)  # shape: (batch, seq_length)
```

Comment:
Will this work with factored models? `batch.target` should be of shape `(batch, target_length, num_target_factors)` and `num_target_factors` might be >1. Probably better to take the 0th slice of the 3rd dim here.

Reply:
This also makes me realize that we've never test this with a factored model. Should probably have done it to ensure there is no other dimension bug like this.